k8s/envoy-opa/README.md

```diff
@@ -0,0 +1,357 @@
+            "x-forwarded-proto": "http",
+            "x-request-id": "e0939bcf-8beb-4910-a980-be0468ec023f"
+          },
+          "method": "GET",
+          "path": "/profiles/2",
+          "protocol": "HTTP/1.1"
+        },
+      ...
+      },
+      ...
+    },
+   ...
+  },
+
+  "msg": "Decision Log",
+  "query": "data.envoy.authz.allow",
+  "requested_by": "",
+  "result": true,
+  "time": "2020-06-11T18:58:47Z",
+  "timestamp": "2020-06-11T18:58:47.840319148Z",
+  "type": "openpolicyagent.org/decision_logs"
+}
+```
+
+The OPA `result` decision is true in this case, meaning the request is allowed to pass through the filter and reach the `backend` service because all of the following conditions defined in the `opa-policy.rego` Rego policy are met:
+* The SPIFFE ID URI extracted from the `x-forwarded-client-cert` (XFCC) header matches the expected SPIFFE ID: `spiffe://example.org/ns/default/sa/default/backend`
+* The request's path matches: `/profiles/2`
+* The HTTP method matches: `GET`
+
+## Testing Invalid Requests
+
+On the other hand, following the same steps we can confirm that a request that does not satisfy the policy prevents the associated data from being displayed. In this case the `frontend-2` service is not able to talk to the `backend` service because its SPIFFE ID does not satisfy the policy for the OPA Agent.
+As a result, when you connect to the URL for the `frontend-2` service (e.g. `http://35.222.190.182:3002`), the browser only displays the title without any account details.
+
+![Frontend-2-no-details][frontend-2-view-no-details]
+
+[frontend-2-view-no-details]: images/frontend-2_view_no_details.png "Frontend-2 view no details account"
+
+After trying to display the `frontend-2` data, you can verify the decision made by OPA using the same `scripts/backend-opa-logs.sh` script as performed in the previous section. A similar log entry is available for the `frontend-2` service but with the result equal to `false` due to the SPIFFE ID mismatch.
+
+## Retesting frontend-2 with a New Policy
+
+Let's update the Rego policy to match the SPIFFE ID of the `frontend-2` and test again. There is a Bash script that we can leverage to complete this task. Once executed, it will open the editor defined by your `KUBE_EDITOR`, or `EDITOR` environment variables, or fall back to `vi` for Linux or Notepad for Windows.
+
+```console
+$ bash scripts/backend-update-policy.sh
+```
+
+With the editor open, look for the following line that specifies the SPIFFE ID:
+
+```console
+svc_spiffe_id == "spiffe://example.org/ns/default/sa/default/frontend"
+```
+
+Update that line to match the SPIFFE ID for the `frontend-2` workload:
+
+```console
+svc_spiffe_id == "spiffe://example.org/ns/default/sa/default/frontend-2"
+```
+
+Save the changes and exit. The `backend-update-policy.sh` script resumes. The script applies the new version of the ConfigMap and then restarts the `backend` pod to pickup the new rule.
```

Comment:
"Pick up" is the verb. "Pickup" is a truck or a sexy meeting with a stranger. But many English speakers don't know this rule either.

```suggestion
Save the changes and exit. The `backend-update-policy.sh` script resumes. The script applies the new version of the ConfigMap and then restarts the `backend` pod to pick up the new rule.
```

Reply:
This is a good catch :100: